File: 502_data mining/module3/data_set_rebalance.py
#balancing training data in python 
#assume some dataset bank_train with our training data 

bank_train['response'].value_counts() #<- identify how many records have the less common value 

#assume our yes count = 338 for n = 3089 and we want the rate of rare to be 30%
#formula: x = 0.3(3089)-338 / 0.7 = 841 
#we need to add 841 records. 

to_resample <- bank.train.loc[bank_train['response']=='yes']
our_resample <- to_resample.sample(n=841,replace=True)
bank_train_rebal <- pd.concat([bank_train,our_resample])

File: 502_data mining/module3/data_set_rebalance.r
#balancing a dataset using R 

table(bank_train$response) #check which variable needs to be rebalanced 
						   #assume n = 3103 with yes = 336 (11%)
						   #assume we want 30% 
						   
#formula:
#x = 0.3(3103)-336 / 0.7 = 850 

to.resample<- which(bank_train$response=='yes') #get row numbers that correspond to cond 
our.resample<- sample(x=to.resample, size=850, replace=TRUE)

our.resample <- bank_train[our.resample,]
train_bank_rebal <- rbind(bank_train,our.resample)

#confirm:
t.v1 <- table(train_bank_rebal$response)
t.v2 <- rbind(t.v1, round(prop.table(t.v1), 4))
colnames(t.v2) <- c("Response = No", "Response = Yes");
rownames(t.v2) <- c("Count", "Proportion")
t.v2
